Create the Scale Applications Using SAP BTP Cloud Logging tutorial (#931)
kyma-project/keda-manager: docs/user/tutorials/01-10-scale-with-cls.md

# Scale Applications Using SAP BTP Cloud Logging Service Metrics

This tutorial shows how to configure KEDA to autoscale a Kubernetes workload using application metrics stored in SAP BTP Cloud Logging Service (CLS) as the scaling signal.

SAP BTP Cloud Logging Service is a managed observability backend built on OpenSearch. When your application metrics already flow into CLS through the [Kyma Telemetry module](https://kyma-project.io/external-content/telemetry-manager/docs/user/README.html), you can use those metrics as autoscaling signals without running a separate metrics store.

KEDA 2.20 ships a native `opensearch` scaler that queries CLS directly using an inline query. This tutorial walks you from a demo app that emits a `queue_depth` metric, through Telemetry scraping, to a `ScaledObject` that scales the workload based on that metric.

## Prerequisites

> - The Keda, Telemetry, and BTP Operator modules are enabled in your Kyma cluster. See [Quick Install](https://kyma-project.io/02-get-started/01-quick-install.html).

Reviewer: ```suggestion
- The Keda, Telemetry, and SAP BTP Operator modules are enabled in your Kyma cluster. See [Quick Install](https://kyma-project.io/02-get-started/01-quick-install.html).
```

- Your subaccount has an entitlement for Cloud Logging Service with the `standard` plan. See [Configure Entitlements and Quotas for Subaccounts](https://help.sap.com/docs/btp/sap-business-technology-platform/configure-entitlements-and-quotas-for-subaccounts).
- `kubectl` is installed and configured to access your Kyma cluster.

## Steps

### Create a CLS Instance and Credentials

<!-- tabs:start -->

> #### **BTP Cockpit**

Reviewer: ```suggestion
#### **SAP BTP cockpit**
```

Reviewer: Pls see the [approved name](https://brand.sap.com/naming-center/requests/10022#23167) 🙈

> Create the CLS instance in a separate subaccount using SAP BTP Cockpit. The instance and its credentials are independent of any Kyma cluster, so they survive cluster replacement or deletion.

Reviewer: ```suggestion
Create the CLS instance in a separate subaccount using the SAP BTP cockpit. The instance and its credentials are independent of any Kyma cluster, so they survive cluster replacement or deletion.
```

> 1. In SAP BTP Cockpit, go to **Services** → **Instances and Subscriptions** and choose **Create**.

Reviewer: ```suggestion
1. In the SAP BTP cockpit, go to **Services** → **Instances and Subscriptions** and choose **Create**.
```

   - **Plan**: standard
   - **Runtime Environment**: Other
   - **Instance Name**: Enter a name, for example `cloud-logging`

3. In the **Parameters** field, add the following parameters to the default JSON and choose **Create**:

    ```json
    {
        "backend": {
            "api_enabled": true,
            "max_data_nodes": 2
        },
        "ingest_otlp": {
            "enabled": true
        }
    }
    ```

4. Wait until the instance status changes to **Created**.

5. In the instance row, choose the **...** (Actions) menu and select **Create Service Binding**. Enter a name for the binding, for example `cloud-logging-binding`, and choose **Create**.

6. Once the binding is created, choose **View Credentials** from the same **...** menu. Note the following values:

   | Key | Description |
   |---|---|
   | `backend-endpoint` | OpenSearch REST API endpoint |
   | `backend-username` | Username for OpenSearch REST API authentication |
   | `backend-password` | Password for OpenSearch REST API authentication |
   | `ingest-otlp-endpoint` | OTLP ingest endpoint for the Telemetry module |
   | `ingest-otlp-cert` | Client certificate for mTLS |

>    | `ingest-otlp-key` | Client key for mTLS |

Reviewer: Shouldn't we use bold font for the keys? Like for parameters, fields, and variables. Also, in line 152.

7. Create a namespace and a Kubernetes Secret with the credentials:

    ```bash
    kubectl create namespace cls
    ```

    ```bash
    kubectl apply -f - <<EOF
    apiVersion: v1
    kind: Secret
    metadata:
      name: cloud-logging-binding
      namespace: cls
    type: Opaque
    stringData:
      backend-endpoint: "<BACKEND_ENDPOINT>"
      backend-username: "<BACKEND_USERNAME>"
      backend-password: "<BACKEND_PASSWORD>"
      ingest-otlp-endpoint: "<INGEST_OTLP_ENDPOINT>"
      ingest-otlp-cert: |
        <INGEST_OTLP_CERT>
      ingest-otlp-key: |
        <INGEST_OTLP_KEY>
    EOF
    ```

>     Replace each placeholder with the corresponding value from the Cockpit. For `ingest-otlp-cert` and `ingest-otlp-key`, paste the full PEM content including the `-----BEGIN ...-----` and `-----END ...-----` lines.

Reviewer: ```suggestion
    Replace each placeholder with the corresponding value from the cockpit. For `ingest-otlp-cert` and `ingest-otlp-key`, paste the full PEM content including the `-----BEGIN ...-----` and `-----END ...-----` lines.
```

#### **SAP BTP Operator module**

Create the CLS instance directly in your Kyma cluster using the SAP BTP Operator module.

> 1. Create a namespace for CLS resources and a `ServiceInstance` for Cloud Logging:

Reviewer: ```suggestion
1. Create a namespace for CLS resources and a service instance for Cloud Logging:
```

Reviewer: At some point, we decided to use natural language for SIs and SBs unless there's a good reason not to, pls decide for yourself.

    ```bash
    kubectl create namespace cls
    ```

    ```bash
    kubectl apply -f - <<EOF
    apiVersion: services.cloud.sap.com/v1
    kind: ServiceInstance
    metadata:
      name: cloud-logging
      namespace: cls
    spec:
      serviceOfferingName: cloud-logging
      servicePlanName: standard
      parameters:
        backend:
          api_enabled: true
          max_data_nodes: 2
        ingest_otlp:
          enabled: true
    EOF
    ```

2. Wait until the instance is ready:

    ```bash
    kubectl get serviceinstance cloud-logging -n cls -w
    ```

    You should get a result similar to this example:

    ```bash
    NAME             OFFERING        PLAN       STATUS    AGE
    cloud-logging    cloud-logging   standard   Created    2m
    ```

> 3. Create a `ServiceBinding` to generate the credentials Secret:

Reviewer: ```suggestion
3. Create a service binding to generate the credentials Secret:
```

    ```bash
    kubectl apply -f - <<EOF
    apiVersion: services.cloud.sap.com/v1
    kind: ServiceBinding
    metadata:
      name: cloud-logging-binding
      namespace: cls
    spec:
      serviceInstanceName: cloud-logging
    EOF
    ```

4. Verify that the binding Secret was created and contains the required keys:

    ```bash
    kubectl get secret cloud-logging-binding -n cls -o jsonpath='{.data}' | jq 'keys'
    ```

    The Secret must contain `backend-endpoint`, `backend-username`, `backend-password`, `ingest-otlp-endpoint`, `ingest-otlp-cert`, and `ingest-otlp-key`.

<!-- tabs:end -->

### Share a CLS Instance Across Clusters (Optional)

If you want to reuse a single CLS instance across multiple Kyma clusters or subaccounts, you can use SAP Service Manager instance sharing. This lets you create a pointer instance in each cluster that references the shared instance, without affecting its lifecycle when the pointer is deleted.

> 1. Mark your CLS instance as sharable. In SAP BTP Cockpit, choose **Share Instance** from the **...** menu for your CLS instance. Alternatively, use the btp CLI:

Reviewer: ```suggestion
1. Mark your CLS instance as sharable. In the SAP BTP cockpit, choose **Share Instance** from the **...** menu for your CLS instance. Alternatively, use the btp CLI:
```

    ```bash
    btp share services/instance <instance-id> --subaccount <subaccount-id>
    ```

> 2. In each Kyma cluster that needs access, create a `ServiceInstance` with the `reference-instance` plan pointing to the shared instance:

Reviewer: ```suggestion
2. In each Kyma cluster that needs access, create a service instance with the `reference-instance` plan pointing to the shared instance:
```

    ```bash
    kubectl apply -f - <<EOF
    apiVersion: services.cloud.sap.com/v1
    kind: ServiceInstance
    metadata:
      name: cloud-logging-pointer
      namespace: cls
    spec:
      serviceOfferingName: cloud-logging
      servicePlanName: reference-instance
      parameters:
        instance_name_selector: "<SHARED_INSTANCE_NAME>"
    EOF
    ```

> 3. Create a `ServiceBinding` on the pointer instance as described in the **SAP BTP Operator module** tab above.

Reviewer: ```suggestion
3. Create a service binding on the pointer instance as described in the **SAP BTP Operator module** tab.
```

      namespaces:
        include:
          - keda-cls-demo
    istio:
      enabled: false
    runtime:
      enabled: false
    otlp:
      enabled: true
  output:
    otlp:
      endpoint:
        valueFrom:
          secretKeyRef:
            name: cloud-logging-binding
            namespace: cls
            key: ingest-otlp-endpoint
      tls:
        cert:
          valueFrom:
            secretKeyRef:
              name: cloud-logging-binding
              namespace: cls
              key: ingest-otlp-cert
        key:
          valueFrom:
            secretKeyRef:
              name: cloud-logging-binding
              namespace: cls
              key: ingest-otlp-key
EOF
```

Verify the pipeline is ready:

```bash
kubectl get metricpipeline cls-metric-pipeline
```

You should get a result similar to this example:

```bash
NAME                  CONFIGURATION GENERATED   GATEWAY HEALTHY   AGENT HEALTHY   FLOW HEALTHY   AGE
cls-metric-pipeline   True                      True              True            True           2m
```

In the CLS OpenSearch Dashboards, confirm that the `queue_depth` metric is arriving:

1. Get the Dashboards URL:

    <!-- tabs:start -->

>     #### **SAP BTP Cockpit**

Reviewer: ```suggestion
    #### **SAP BTP cockpit**
```

    Find the `dashboards-endpoint` value under **View Credentials** for your CLS binding.

    #### **SAP BTP Operator module**

    Run the following command:

    ```bash
    echo "https://$(kubectl get secret cloud-logging-binding -n cls -o jsonpath='{.data.dashboards-endpoint}' | base64 -d)"
    ```

    <!-- tabs:end -->

2. Open the URL in your browser and log in with the Dashboards credentials from your CLS binding.

3. In the navigation menu, go to **Discover**, select the `metrics-otel-v1-*` index pattern, and filter for documents with `name: queue_depth`. The metric should appear within 1-2 minutes after the MetricPipeline becomes healthy.

### Create a KEDA TriggerAuthentication for CLS

KEDA must authenticate with the CLS OpenSearch REST API to run queries. Store the CLS credentials in a Kubernetes `Secret` and reference them from a `TriggerAuthentication`.

The credentials are available in the CLS service binding Secret. The following keys are used:

| Key | Description |
|---|---|
| `backend-endpoint` | OpenSearch REST API endpoint |
| `backend-username` | Username for OpenSearch REST API authentication |

> | `backend-password` | Password for OpenSearch REST API authentication |

Reviewer: Bold font?

> 1. Create a `Secret` with your CLS OpenSearch credentials:

Reviewer: ```suggestion
1. Create a Secret with your CLS OpenSearch credentials:
```

    ```bash
    kubectl set env deployment/fake-metrics QUEUE_DEPTH=80 -n keda-cls-demo
    kubectl rollout restart deployment/fake-metrics -n keda-cls-demo
    ```

    After the next Telemetry scrape and CLS ingestion cycle (typically within 1-2 minutes), KEDA queries CLS and adjusts the replica count.

3. Watch the Pods scale up:

    ```bash
    kubectl get pods -n keda-cls-demo -w
    ```

4. Set the metric back to a lower value to observe scale-down:

    ```bash
    kubectl set env deployment/fake-metrics QUEUE_DEPTH=5 -n keda-cls-demo
    kubectl rollout restart deployment/fake-metrics -n keda-cls-demo
    ```

    After the cooldown period, the replica count returns to the minimum of 1. This may take up to 5 minutes.

## Result

Your workload scales automatically in response to the `queue_depth` metric stored in SAP BTP Cloud Logging Service. KEDA polls CLS on each reconciliation cycle and adjusts the HPA target accordingly.

## Clean Up

If you want to remove the resources created during this tutorial, run the following commands:

```bash
kubectl delete namespace keda-cls-demo
kubectl delete metricpipeline cls-metric-pipeline
kubectl delete namespace cls
```

> If you used the BTP Operator tab, also delete the service binding and instance:

Reviewer: ```suggestion
If you used the SAP BTP Operator tab, also delete the service binding and instance:
```

```bash
kubectl delete servicebinding cloud-logging-binding -n cls
kubectl delete serviceinstance cloud-logging -n cls
```

> If you used the BTP Cockpit tab, delete the CLS instance in BTP Cockpit under **Services** → **Instances and Subscriptions**.

Reviewer: ```suggestion
If you used the SAP BTP cockpit tab, delete the CLS instance in the cockpit under **Services** → **Instances and Subscriptions**.
```
